### PR TITLE
add bench

### PR DESCRIPTION
**Pull request by 6293**

### CODE REVIEW

**Review by dr-orlovsky — APPROVED** [review]
Why draft?

**Comment by 6293** [comment]
@dr-orlovsky CI is failling. fixing it

File: Cargo.toml [hunk 1]
```diff
@@ -32,6 +32,11 @@ serde_yaml = { version = "0.8", optional = true }
 toml = { version = "0.5", optional = true }
 stringly_conversions = { version = "0.1.1", optional = true, features = ["alloc"] }
 
+# avoid building criterion in 1.41.1 CI
+[target.'cfg(bench)'.dev-dependencies]
+criterion = "0.2.11"
+softposit = "0.3.9"
```
Comment (6293) [inline Cargo.toml:38]: workaround when one wants to run criterion while supporting old rust compiler

https://github.com/rust-lang/cargo/issues/1596#issuecomment-962730780

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -50,13 +50,13 @@ jobs:
         uses: actions-rs/cargo@v1
         with:
           command: check
-          args: --workspace --all-targets --all-features
+          args: --workspace --lib --bins --tests --examples --all-features
```
Comment (6293) [inline .github/workflows/build.yml:53]: avoid building `benches` directory

**Comment by 6293** [comment]
### Result

```
u1024_add               time:   [19.950 ns 19.974 ns 19.997 ns]                       
Found 3 outliers among 100 measurements (3.00%)
  1 (1.00%) low mild
  2 (2.00%) high mild

u1024_sub               time:   [25.290 ns 25.859 ns 26.645 ns]                       
Found 20 outliers among 100 measurements (20.00%)
  6 (6.00%) high mild
  14 (14.00%) high severe

u1024_mul               time:   [225.26 ns 225.55 ns 225.85 ns]                      
Found 3 outliers among 100 measurements (3.00%)
  3 (3.00%) high mild

u1024_div               time:   [135.14 ns 135.23 ns 135.33 ns]                      
Found 12 outliers among 100 measurements (12.00%)
  6 (6.00%) high mild
  6 (6.00%) high severe

posit32_add             time:   [14.012 ns 14.032 ns 14.056 ns]                         
Found 6 outliers among 100 measurements (6.00%)
  5 (5.00%) high mild
  1 (1.00%) high severe

posit32_sub             time:   [14.707 ns 14.745 ns 14.802 ns]                         
Found 6 outliers among 100 measurements (6.00%)
  5 (5.00%) high mild
  1 (1.00%) high severe

posit32_mul             time:   [12.327 ns 12.341 ns 12.355 ns]                         
Found 5 outliers among 100 measurements (5.00%)
  3 (3.00%) high mild
  2 (2.00%) high severe

posit32_div             time:   [15.108 ns 15.137 ns 15.175 ns]                         
Found 12 outliers among 100 measurements (12.00%)
  4 (4.00%) high mild
  8 (8.00%) high severe

softposit32_add         time:   [4.6523 ns 4.6600 ns 4.6676 ns]                             
Found 5 outliers among 100 measurements (5.00%)
  4 (4.00%) high mild
  1 (1.00%) high severe

softposit32_sub         time:   [5.1498 ns 5.1584 ns 5.1688 ns]                             
Found 13 outliers among 100 measurements (13.00%)
  2 (2.00%) high mild
  11 (11.00%) high severe

softposit32_mul         time:   [4.9127 ns 4.9281 ns 4.9459 ns]                             
Found 8 outliers among 100 measurements (8.00%)
  6 (6.00%) high mild
  2 (2.00%) high severe

softposit32_div         time:   [6.9134 ns 6.9266 ns 6.9405 ns]                             
Found 8 outliers among 100 measurements (8.00%)
  4 (4.00%) high mild
  4 (4.00%) high severe
```

**Comment by 6293** [comment]
[softposit](https://gitlab.com/burrbull/softposit-rs) is our predecessor which no longer actively developed. Currently we are 3x slower than them because I have not been caring about optimization, and we could outperform them in the future

File: .github/workflows/build.yml (hunk 2 above)
Comment (dr-orlovsky) [reply .github/workflows/build.yml:53]: That will result in a fact that future PRs can break benches and we will not notice. Maybe add a separate CI to not run benches on 1.41.1 but still run on other versions?

**Review by dr-orlovsky — COMMENTED** [review]
Pls see my comment

File: .github/workflows/build.yml (hunk 2 above)
Comment (6293) [reply .github/workflows/build.yml]: that's true. fixed

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -48,6 +48,8 @@ jobs:
       - name: All features
         if: matrix.toolchain != '1.41.1'
         uses: actions-rs/cargo@v1
+        env:
+          RUSTFLAGS: "--cfg bench"
```
Comment (dr-orlovsky) [inline .github/workflows/build.yml:52]: Sorry for not being explicit; what I meant was to have if-test in github action checking for the compiler version and than skipping/running bench tests depending on the result

Comment (6293) [reply .github/workflows/build.yml:52]: I am sorry that I don't see well what you mean. Is `if: matrix.toolchain != '1.41.1'` not sufficient for `checking for the compiler version`?

Comment (6293) [reply .github/workflows/build.yml:52]: cargo downloads and compiles `criterion` if and only if `RUSTFLAGS="--cfg bench"`. So we have to specify `cfg` to use `--all-targets`

Comment (dr-orlovsky) [reply .github/workflows/build.yml:52]: Yes, the matrix if check is sufficient. But I do not get why we need to explicitly specify `cfg` if it should be automatically assigned by `cargo bench`

Comment (6293) [reply .github/workflows/build.yml:52]: > if it should be automatically assigned by cargo bench

It is not assigned by cargo. `test` is the only cfg they assign

**Review by dr-orlovsky — APPROVED** [review]
ACK

**Pull request merged**